### PR TITLE
LWDEV-10416 Use log4j2 in ME

### PR DESCRIPTION
**Pull request by rocketmannot**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -45,19 +44,21 @@ ext {
...
+    lmax_disruptor_version = '3.4.2'
     springfox_swagger2 = "2.9.2"
     springfox_swagger_ui = "2.9.2"
 }
 
 dependencies {
     // internal libs
-    compile "org.bitbucket.lykke:kotlin-utils:$kotlin_utils_version"
+    compile ("org.bitbucket.lykke:kotlin-utils:$kotlin_utils_version") {
+        exclude module: 'log4j'
```
Comment (barsukovskii) [inline build.gradle]: As I understood this exclusion is not mandatory as log4j dependency has been removed from utils lib

File: src/dist/cfg/log4j2.xml [hunk 2]
```diff
@@ -0,0 +1,653 @@
+<Configuration>
+    <Appenders>
+        <!-- Async Loggers will auto-flush in batches, so switch off immediateFlush. -->
```
Comment (barsukovskii) [inline src/dist/cfg/log4j2.xml]: Seems like the comment is not actual as actually immediateFlush=true

File: src/dist/cfg/log4j2.xml [hunk 3]
```diff
@@ -0,0 +1,653 @@
...
+                                 immediateFlush="true">
+            <PatternLayout>
+                <Pattern>%d{dd-MM HH:mm:ss:SSS} %p [%t] [%c{1}] %m%n</Pattern>
+            </PatternLayout>
+            <Policies>
+                <TimeBasedTriggeringPolicy interval="1" modulate="true"/>
+            </Policies>
+        </RollingRandomAccessFile>
+
+        <RollingRandomAccessFile name="ReservedBalanceLog" filePattern="../log/reservedBalanceUpdate.log.%d{yyyy-MM-dd}"
```
Comment (barsukovskii) [inline src/dist/cfg/log4j2.xml]: outdated appender, ReservedBalanceUpdateService doesn't exist

File: src/dist/cfg/log4j2.xml [hunk 4]
```diff
@@ -0,0 +1,653 @@
...
+        </AsyncLogger>
+
+        # HttpClient for metrics
+        <AsyncLogger name="org.apache.http"
+                     level="info" additivity="false">
+            <AppenderRef ref="ServerLog"/>
+        </AsyncLogger>
+
+        # Reserved Balance Update
+        <AsyncLogger name="com.lykke.matching.engine.services.ReservedBalanceUpdateService"
```
Comment (barsukovskii) [inline src/dist/cfg/log4j2.xml]: outdated logger, ReservedBalanceUpdateService doesn't exist

File: src/dist/cfg/log4j2.xml [hunk 5]
```diff
@@ -0,0 +1,653 @@
...
+                                 immediateFlush="true">
+            <PatternLayout>
+                <Pattern>%d{dd-MM HH:mm:ss:SSS} %p [%c{1}] %m%n</Pattern>
+            </PatternLayout>
+            <Policies>
+                <TimeBasedTriggeringPolicy interval="1" modulate="true"/>
+            </Policies>
+        </RollingRandomAccessFile>
+
+        <RollingRandomAccessFile name="SingleLimitStatsLog" filePattern="../log/singleLimitOrder.log.%d{yyyy-MM-dd}"
```
Comment (barsukovskii) [inline src/dist/cfg/log4j2.xml]: singleLimitOrder -> statsLimitOrder, and at line 82

File: src/dist/cfg/log4j2.xml [hunk 6]
```diff
@@ -0,0 +1,653 @@
...
+                                 immediateFlush="true">
+            <PatternLayout>
+                <Pattern>%d{dd-MM HH:mm:ss:SSS} %p [%c{1}] %m%n</Pattern>
+            </PatternLayout>
+            <Policies>
+                <TimeBasedTriggeringPolicy interval="1" modulate="true"/>
+            </Policies>
+        </RollingRandomAccessFile>
+
+        <RollingRandomAccessFile name="GenericLimitOrderLog" filePattern="../log/genericLimitOrder.log.%d{yyyy-MM-dd}"
```
Comment (barsukovskii) [inline src/dist/cfg/log4j2.xml]: outdated appender

File: src/dist/cfg/log4j2.xml [hunk 7]
```diff
@@ -0,0 +1,653 @@
...
+        </AsyncLogger>
+
+        # Multi Limit Cancel Order
+        <AsyncLogger name="com.lykke.matching.engine.services.MultiLimitOrderCancelService"
+                     level="debug" additivity="false">
+            <AppenderRef ref="MultiLimitCancelLog"/>
+        </AsyncLogger>
+
+        # Generic Limit Order
+        <AsyncLogger name="com.lykke.matching.engine.services.GenericLimitOrderService"
```
Comment (barsukovskii) [inline src/dist/cfg/log4j2.xml]: outdated logger

File: src/dist/cfg/log4j2.xml [hunk 8]
```diff
@@ -0,0 +1,653 @@
...
+                                 immediateFlush="true">
+            <PatternLayout>
+                <Pattern>%d{dd-MM HH:mm:ss:SSS} %p [%c{1}] %m%n</Pattern>
+            </PatternLayout>
+            <Policies>
+                <TimeBasedTriggeringPolicy interval="1" modulate="true"/>
+            </Policies>
+        </RollingRandomAccessFile>
+
+        <RollingRandomAccessFile name="WalletCredentialsLog"
```
Comment (barsukovskii) [inline src/dist/cfg/log4j2.xml]: outdated appender, WalletCredentialsCache doesn't exist

File: src/dist/cfg/log4j2.xml [hunk 9]
```diff
@@ -0,0 +1,653 @@
...
+        </AsyncLogger>
+
+        # Queue size logger
+        <AsyncLogger name="com.lykke.matching.engine.utils.QueueSizeLogger"
+                     level="debug" additivity="false">
+            <AppenderRef ref="QueueSizeLog"/>
+        </AsyncLogger>
+
+        # Wallet Cache
+        <AsyncLogger name="com.lykke.matching.engine.cache.WalletCredentialsCache"
```
Comment (barsukovskii) [inline src/dist/cfg/log4j2.xml]: outdated logger, WalletCredentialsCache doesn't exist

File: src/dist/cfg/log4j2.xml [hunk 10]
```diff
@@ -0,0 +1,653 @@
...
+                <Pattern>%d{dd-MM HH:mm:ss:SSS} %p [%t] [%c{1}] %m%n</Pattern>
+            </PatternLayout>
+        </Console>
+    </Appenders>
+    <Loggers>
+        <Root level="debug" additivity="false">
+            <AppenderRef ref="ServerLog"/>
+        </Root>
+        <!-- pattern layout actually uses location, so we need to include it -->
+        <AsyncLogger name="AppStarter" level="info" additivity="false">
```
Comment (barsukovskii) [inline src/dist/cfg/log4j2.xml]: Propose to leave all loggers which are used while starting ME a synchronous

File: src/dist/cfg/log4j2.xml [hunk 11]
```diff
@@ -0,0 +1,653 @@
...
+            <AppenderRef ref="CashLog"/>
+        </AsyncLogger>
+        `
+        # Limit Order Cancel
+        <Logger name="LimitOrderCancelPreProcessing"
+                level="debug" additivity="false">
+            <AppenderRef ref="LimitCancelLog"/>
+        </Logger>
+
+        <Logger name="LimitOrderCancelPreProcessing"
```
Comment (barsukovskii) [inline src/dist/cfg/log4j2.xml]: LimitOrderCancelPreProcessing -> LimitOrderMassCancelPreProcessing

File: src/dist/cfg/log4j2.xml [hunk 12]
```diff
@@ -0,0 +1,653 @@
...
+        </AsyncLogger>
+
+        # History ticks persist stats
+        <AsyncLogger name="marketStateCache"
+                     level="debug" additivity="false">
+            <AppenderRef ref="MarketStateCacheLog"/>
+        </AsyncLogger>
+
+        # Order book requests
+        <AsyncLogger name="com.lykke.matching.engine.web.controllers.web.OrderBooksController"
```
Comment (barsukovskii) [inline src/dist/cfg/log4j2.xml]: wrong package name (com.lykke.matching.engine.web.controllers.web -> com.lykke.matching.engine.web.controllers)

File: build.gradle (hunk 1 above)
Comment (rocketmannot) [reply build.gradle]: fixed

File: src/dist/cfg/log4j2.xml (hunk 2 above)
Comment (rocketmannot) [reply src/dist/cfg/log4j2.xml]: fixed

File: src/dist/cfg/log4j2.xml (hunk 3 above)
Comment (rocketmannot) [reply src/dist/cfg/log4j2.xml]: fixed

File: src/dist/cfg/log4j2.xml (hunk 4 above)
Comment (rocketmannot) [reply src/dist/cfg/log4j2.xml]: fixed

File: src/dist/cfg/log4j2.xml (hunk 12 above)
Comment (rocketmannot) [reply src/dist/cfg/log4j2.xml]: fixed

File: src/dist/cfg/log4j2.xml (hunk 11 above)
Comment (rocketmannot) [reply src/dist/cfg/log4j2.xml]: fixed

File: src/dist/cfg/log4j2.xml (hunk 10 above)
Comment (rocketmannot) [reply src/dist/cfg/log4j2.xml]: agree

File: src/dist/cfg/log4j2.xml (hunk 9 above)
Comment (rocketmannot) [reply src/dist/cfg/log4j2.xml]: fixed

File: src/dist/cfg/log4j2.xml (hunk 8 above)
Comment (rocketmannot) [reply src/dist/cfg/log4j2.xml]: fixed

File: src/dist/cfg/log4j2.xml (hunk 5 above)
Comment (rocketmannot) [reply src/dist/cfg/log4j2.xml]: fixed

File: src/dist/cfg/log4j2.xml (hunk 6 above)
Comment (rocketmannot) [reply src/dist/cfg/log4j2.xml]: fixed

File: src/dist/cfg/log4j2.xml (hunk 7 above)
Comment (rocketmannot) [reply src/dist/cfg/log4j2.xml]: fixed

File: src/dist/cfg/log4j2.xml (hunk 5 above)
Comment (rocketmannot) [reply src/dist/cfg/log4j2.xml]: fixed

File: src/dist/cfg/log4j2.xml [hunk 13]
```diff
@@ -0,0 +1,618 @@
...
+                <Pattern>%d{dd-MM HH:mm:ss:SSS} %p [%c{1}] %m%n</Pattern>
+            </PatternLayout>
+            <Policies>
+                <TimeBasedTriggeringPolicy interval="1" modulate="true"/>
+            </Policies>
+        </RollingRandomAccessFile>
+
+        <RollingRandomAccessFile name="RabbitMqPublisherStatsLog"
+                                 filePattern="../log/statsRabbitMessages.log.%d{yyyy-MM-dd}"
+                                 fileName=".../log/statsRabbitMessages.log"
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/dist/cfg/log4j2.xml]: fixed

File: src/dist/bin/start.bat [hunk 14]
```diff
@@ -3,6 +3,6 @@ set DIRNAME=%~dp0
 if "%DIRNAME%" == "" set DIRNAME=.
 set APP_HOME=%DIRNAME%..
 
-set LYKKE_ME_PROTOTYPE_OPTS="-Dlog4j.configuration=file:///%APP_HOME%/cfg/log4j.properties"
+set LYKKE_ME_PROTOTYPE_OPTS="-Dlog4j.configuration=file:///%APP_HOME%/cfg/log4j2.xml"
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/dist/bin/start.bat]: fixed

File: src/dist/cfg/log4j2.xml [hunk 15]
```diff
@@ -0,0 +1,618 @@
...
+        <Root level="debug" additivity="false">
+            <AppenderRef ref="ServerLog"/>
+        </Root>
+
+        <Logger name="AppStarter" level="info" additivity="false">
+            <AppenderRef ref="ServerLog"/>
+            <AppenderRef ref="ConsoleLog"/>
+        </Logger>
+
+        <AsyncLogger name="com.lykke.matching.engine.utils.order.MinVolumeOrderCanceller"
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/dist/cfg/log4j2.xml]: fixed

File: src/dist/cfg/log4j2.xml [hunk 16]
```diff
@@ -0,0 +1,618 @@
...
+                     level="info" additivity="false">
+            <AppenderRef ref="ServerLog"/>
+            <AppenderRef ref="ConsoleLog"/>
+
+        </AsyncLogger>
+        <AsyncLogger name="com.lykke.matching.engine.utils.order.AllOrdersCanceller"
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/dist/cfg/log4j2.xml]: fixed

**Pull request merged**